### ISSUE TITLE
See if we can provide a single API for construcing Cron expressions

### ISSUE THREAD

**Issue by rix0rrr** [aws/aws-cdk#1181]
Currently, CloudWatch Events, Application AutoScaling and AutoScaling all have cron expressions, and so they have a `Cron` helper class. See if we can merge these, and make a more expressive cron expression builder.

At least the syntaxes of AppAutoScaling and AutoScaling are not compatible, they look like this:

```
cron(0 18 * * ?)    # appscaling
0 18 * * *             # autoscaling
```

Note the cron() function, and the difference between * and ? for the DOW field.

See comments in https://github.com/awslabs/aws-cdk/pull/1134

**Comment by rix0rrr** [comment]
Noticed someone struggling on the Gitter channel, putting in here for reference:

```
there is something funny about eventRule scheduleExpressions....

this cdk code:

const rule = new events.EventRule(this, baseName + 'ScheduledEvent', {
    scheduleExpression: 'cron(15 0 * * MON-FRI *)', // 00:15 every weekday
})

produces a "Parameter ScheduleExpression is not valid." failure during deploy

but this code:

const rule = new events.EventRule(this, baseName + 'ScheduledEvent', {
    scheduleExpression: 'cron(15 0 * * 2-6 *)', // 00:15 every weekday
})

works fine. 'MON-FRI' should be a valid field for Cloudwatch. Is this a cdk issue or a cfn issue?
oh i take it back. neither work :(

cdk synth shows this: ScheduleExpression: cron(15 0 * * 2-6 *)
only this works: scheduleExpression: 'cron(15 0 * * ? *)' :(

oops my bad: cron(15 0 ? * MON-FRI *) is just fine (needed a '?' in there)
```

**Comment by wryfi** [comment]
There is also a pretty substantial oversight in [aws-autoscaling/lib/scheduled-action.ts](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-autoscaling/lib/scheduled-action.ts#L78), where the constant that validates cron expressions does not allow commas or dashes:

```
const CRON_PART = '(\\*|\\?|[0-9]+)';

const CRON_EXPRESSION = new RegExp('^' + [CRON_PART, CRON_PART, CRON_PART, CRON_PART, CRON_PART].join('\\s+') + '$');
```

This effectively makes it impossible to create certain types of ASG schedules, like those that do something every weekday, for example.

**Comment by eladb** [comment]
Yap. That’s bad.

**Comment by eladb** [comment]
@wryfi Can you raise a separate issue for this please?

**Comment by wryfi** [comment]
@eladb #2797

**Comment by made2591** [comment]
hi @wryfi @eladb I solved #2797 with a something like

```
const CRON_PART = '(\\*|\\?|\\d+((\\/|\\-){0,1}(\\d+))*)';
```

but, looking at the doc [here](https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-scheduled-scaling.html) it seems you can actually specify a cron expression (`extended`) so... I think that the `CRON_PART` definition should be split into 6 constants with values in respective supported ranges and accepted wildcards as specified [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html#CronExpressions)...

**Comment by NetaNir** [comment]
Since this issue was opened we have added  `cron` method to the  various packages, see [autoscaling](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-autoscaling/lib/schedule.ts#L18), [aws-events](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-events/lib/schedule.ts#L33) and  [applicationautoscaling](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-applicationautoscaling/lib/schedule.ts#L40).
Closing this.
